src/main/java/htsjdk/samtools/util/output/async/AsyncBlockCompressedOutputStream.java

```diff
@@ -0,0 +1,317 @@
+     *                         Note: this constructor uses the default {@link DeflaterFactory}, see {@link #getDefaultDeflaterFactory()}.
+     *                         Use {@link #AsyncBlockCompressedOutputStream(File, int, DeflaterFactory)} to specify a custom factory.
+     */
+    public AsyncBlockCompressedOutputStream(final File file, final int compressionLevel) {
+        this(file, compressionLevel, defaultDeflaterFactory);
+    }
+
+    /**
+     * Prepare to compress at the given compression level
+     *
+     * @param compressionLevel 1 <= compressionLevel <= 9
+     * @param deflaterFactory  custom factory to create deflaters (overrides the default)
+     */
+    public AsyncBlockCompressedOutputStream(final File file, final int compressionLevel, final DeflaterFactory deflaterFactory) {
+        this(IOUtil.toPath(file), compressionLevel, deflaterFactory);
+    }
+
+    public AsyncBlockCompressedOutputStream(final OutputStream os, final Path file) {
+        super(os, file);
+        this.compressionLevel = defaultCompressionLevel;
+        this.deflaterFactory = defaultDeflaterFactory;
+    }
+
+    /**
+     * Note: this constructor uses the default {@link DeflaterFactory}, see {@link #getDefaultDeflaterFactory()}.
+     * Use {@link #AsyncBlockCompressedOutputStream(OutputStream, Path, int, DeflaterFactory)} to specify a custom factory.
+     */
+    public AsyncBlockCompressedOutputStream(final OutputStream os, final Path file, final int compressionLevel) {
+        this(os, file, compressionLevel, defaultDeflaterFactory);
+    }
+
+    /**
+     * Prepare to compress at the given compression level
+     *
+     * @param compressionLevel 1 <= compressionLevel <= 9
+     * @param deflaterFactory  custom factory to create deflaters (overrides the default)
+     */
+    public AsyncBlockCompressedOutputStream(final Path path, final int compressionLevel, final DeflaterFactory deflaterFactory) {
+        super(path);
+        this.compressionLevel = compressionLevel;
+        this.deflaterFactory = deflaterFactory;
+    }
+
+    /**
+     * Creates the output stream.
+     *
+     * @param os               output stream to create a BlockCompressedOutputStream from
+     * @param file             file to which to write the output or null if not available
+     * @param compressionLevel the compression level (0-9)
+     * @param deflaterFactory  custom factory to create deflaters (overrides the default)
+     */
+    public AsyncBlockCompressedOutputStream(final OutputStream os, final Path file, final int compressionLevel, final DeflaterFactory deflaterFactory) {
+        super(os, file);
+        this.compressionLevel = compressionLevel;
+        this.deflaterFactory = deflaterFactory;
+    }
+
+    private static int countCompressingThread() {
+        int availableProcessors = Runtime.getRuntime().availableProcessors();
+        return (availableProcessors <= 2) ? 2 : (availableProcessors - 2);
```

Comment:
- given availableProcessors = 1 then return 2
- given availableProcessors = 2 then return 2
- given availableProcessors = 3 then return 1 // is it designed 
- given availableProcessors = 4 then return 2
- given availableProcessors = 5 then return 3